### PR TITLE
8339288: Improve diagnostic logging runtime/cds/DeterministicDump.java

### PR DESCRIPTION
**Pull request by matias9927**
The test `DeterministicDump.java` fails frequently since it is sensitive to changes in CDS, and with it's current logging, it is very difficult to diagnose the root cause of these failures. The current output looks like this:

```
[First diff: map file #1 (SharedArchiveFile0.map)]
[ro region          0x00000008005a2000 - 0x0000000800ea7058   9457752 bytes]
 [ro region          0x00000008005a2000 - 0x0000000800ea7058   9457752 bytes]
 0x00000008005a2000: @@ Symbol            40 [Ljdk/internal/vm/FillerElement;
>0x00000008005a2000:   4c5b0020e72fffff 65746e692f6b646a 2f6d762f6c616e72 6c4572656c6c6946   ../. .[Ljdk/internal/vm/FillerEl
 0x00000008005a2020:   00003b746e656d65                                                      ement;..
 0x00000008005a2028: @@ Symbol            8 [Z

[First diff: map file #2 (SharedArchiveFile1.map)]
[ro region          0x00000008005a2000 - 0x0000000800ea7058   9457752 bytes]
 [ro region          0x00000008005a2000 - 0x0000000800ea7058   9457752 bytes]
 0x00000008005a2000: @@ Symbol            40 [Ljdk/internal/vm/FillerElement;
>0x00000008005a2000:   4c5b00205d54ffff 65746e692f6b646a 2f6d762f6c616e72 6c4572656c6c6946   ..T] .[Ljdk/internal/vm/FillerEl
 0x00000008005a2020:   00003b746e656d65                                                      ement;..
 0x00000008005a2028: @@ Symbol            8 [Z

Byte #5910530 at line #235268 word #0:
SharedArchiveFile0.map: 4c5b0020e72fffff 
SharedArchiveFile1.map: 4c5b00205d54ffff 
```

This change was verified locally by generating non-deterministic archives and map files.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339288](https://bugs.openjdk.org/browse/JDK-8339288): Improve diagnostic logging runtime/cds/DeterministicDump.java (**Sub-task** - P4)

### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) Review applies to [1ba7318b](https://git.openjdk.org/jdk/pull/21913/files/1ba7318b035a6c98dbc1b68de38bd14d7d225b3d)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21913/head:pull/21913` \
`$ git checkout pull/21913`

Update a local copy of the PR: \
`$ git checkout pull/21913` \
`$ git pull https://git.openjdk.org/jdk.git pull/21913/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21913`

View PR using the GUI difftool: \
`$ git pr show -t 21913`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21913.diff">https://git.openjdk.org/jdk/pull/21913.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21913#issuecomment-2457999229)
</details>

### CODE REVIEW

**Comment by iklam** [comment]
Two suggestions:

(1) I think it will be good to print an indicator on the line that differs. Something like
```
 [First diff: map file #2 (SharedArchiveFile1.map)]
 [ro region          0x00000008005a2000 - 0x0000000800ea7058   9457752 bytes] 
 0x00000008005a2cb0:   6f66000ae5fcffff 6c61766f6d655272                                     ......forRemoval
 0x00000008005a2cc0: @@ Symbol            32 java/lang/VersionProps
 0x00000008005a2cc0:   616a00169955ffff 2f676e616c2f6176 506e6f6973726556 0000000073706f72   ..U...java/lang/VersionProps....
 0x00000008005a2ce0: @@ Symbol            24 java_version
>0x00000008005a2ce0:   616a000ca147ffff 69737265765f6176 0000000000006e6f                    ..G...java_version......
```

(2) Sometimes an object spans over many lines. It will be good to remember the first line of the very last object (the `0x00000008005a2ce0: @@ Symbol` in your example). If that line is more than 4 lines above the first diff, then print something like this:

```
 0x00000008005a2ce0: @@ Symbol            24 java_version
 [....]
```

before printing the rest of the output.

**Comment by matias9927** [comment]
> Two suggestions:
> 
> (1) I think it will be good to print an indicator on the line that differs. Something like
> 
> ```
>  [First diff: map file #2 (SharedArchiveFile1.map)]
>  [ro region          0x00000008005a2000 - 0x0000000800ea7058   9457752 bytes] 
>  0x00000008005a2cb0:   6f66000ae5fcffff 6c61766f6d655272                                     ......forRemoval
>  0x00000008005a2cc0: @@ Symbol            32 java/lang/VersionProps
>  0x00000008005a2cc0:   616a00169955ffff 2f676e616c2f6176 506e6f6973726556 0000000073706f72   ..U...java/lang/VersionProps....
>  0x00000008005a2ce0: @@ Symbol            24 java_version
> >0x00000008005a2ce0:   616a000ca147ffff 69737265765f6176 0000000000006e6f                    ..G...java_version......
> ```
> 
> (2) Sometimes an object spans over many lines. It will be good to remember the first line of the very last object (the `0x00000008005a2ce0: @@ Symbol` in your example). If that line is more than 4 lines above the first diff, then print something like this:
> 
> ```
>  0x00000008005a2ce0: @@ Symbol            24 java_version
>  [....]
> ```
> 
> before printing the rest of the output.

That's a good idea! I think my example output is a little confusing because there are actually two differences in it. Using your idea, the output would look like this:
[First diff: map file #1 (SharedArchiveFile0.map)]
[ro region          0x00000008005a2000 - 0x0000000800ea7058   9457752 bytes]
 0x00000008005a2cb0:   6f66000a8da9ffff 6c61766f6d655272                                     ......forRemoval 
 0x00000008005a2cc0: @@ Symbol            32 java/lang/VersionProps
 >0x00000008005a2cc0:   616a0016959bffff 2f676e616c2f6176 506e6f6973726556 0000000073706f72   ......java/lang/ VersionProps....
 0x00000008005a2ce0: @@ Symbol            24 java_version
 0x00000008005a2ce0:   616a000c335cffff 69737265765f6176 0000000000006e6f                    ..\3..java_version......

[First diff: map file #2 (SharedArchiveFile1.map)]
[ro region          0x00000008005a2000 - 0x0000000800ea7058   9457752 bytes]
 0x00000008005a2cb0:   6f66000ae5fcffff 6c61766f6d655272                                     ......forRemoval
 0x00000008005a2cc0: @@ Symbol            32 java/lang/VersionProps
 >0x00000008005a2cc0:   616a00169955ffff 2f676e616c2f6176 506e6f6973726556 0000000073706f72   ..U...java/lang/ VersionProps....
 0x00000008005a2ce0: @@ Symbol            24 java_version
 0x00000008005a2ce0:   616a000ca147ffff 69737265765f6176 0000000000006e6f                    ..G...java_version......

Byte #5910530 at line #235510 word #0:
SharedArchiveFile0.map: 616a0016959bffff 
SharedArchiveFile1.map: 616a00169955ffff

**Review by iklam — APPROVED** [review]
LGTM.

File: test/lib/jdk/test/lib/cds/CDSArchiveUtils.java [hunk 1]
```diff
@@ -494,6 +495,12 @@ public static File deleteBytesAtTheEnd(File orgFile, String newFileName) throws
...
 
+    // used region size
+    public static long usedRegionSize(File archiveFile, int region) throws Exception {
+        long offset = spOffset + cdsFileMapRegionSize * region + spUsedOffset;
+        return readInt(archiveFile, offset, sizetSize);
+    }
+
     // used region size
     public static long usedRegionSizeAligned(File archiveFile, int region) throws Exception {
         long offset = spOffset + cdsFileMapRegionSize * region + spUsedOffset;
```
Comment (calvinccheung) [inline test/lib/jdk/test/lib/cds/CDSArchiveUtils.java:506]: Can you just call `usedRegionSize` like the following?
`long used = usedRegionSize(archiveFile, region);`

**Review by calvinccheung — APPROVED** [review]
Looks good. Just one nit.

**Review by calvinccheung — APPROVED** [review]

**Comment by matias9927** [comment]
Thanks for the feedback and reviews @iklam and @calvinccheung !
/integrate

**Pull request closed without merging**